File: cli/internal/store/store.go
// Package store locates the project root and reads/writes .incomm/notes[_<branch>].json.
package store

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"incomm/internal/git"
	"incomm/internal/model"
)

// DirName is the directory holding incomm state, at the project root.
const DirName = ".incomm"

// FileName is the default notes file inside DirName (used when no git branch
// is detected or the repo is in detached-HEAD state).
const FileName = "notes.json"

// Store is bound to a resolved project root and an optional git branch.
type Store struct {
	Root      string // absolute path of the directory that holds (or will hold) .incomm/
	RawBranch string // raw git branch name (e.g. "feature/thing"); written into JSON
	Branch    string // sanitized for filename (e.g. "feature_thing"); empty = legacy notes.json
}

// Open resolves the project root and returns a Store.
//
// If explicitRoot is non-empty it is used as the starting point; otherwise the
// current working directory is used. From there, Open walks up the directory
// tree looking for an existing .incomm/ directory and, if found, roots the
// store there. If none is found, the starting directory becomes the root and
// .incomm/ will be created on the first Save.
//
// If explicitBranch is non-empty it is used as the branch name (sanitized for
// filenames). Otherwise the branch is auto-detected by reading .git/HEAD. If
// git is unavailable or the repo is in detached-HEAD state, Branch stays empty
// and the legacy notes.json is used.
func Open(explicitRoot, explicitBranch string) (*Store, error) {
	base := explicitRoot
	if base == "" {
		wd, err := os.Getwd()
		if err != nil {
			return nil, fmt.Errorf("determine working directory: %w", err)
		}
		base = wd
	}
	abs, err := filepath.Abs(base)
	if err != nil {
		return nil, fmt.Errorf("resolve %q: %w", base, err)
	}
	root := abs
	if found, ok := findExisting(abs); ok {
		root = found
	}
	rawBranch := explicitBranch
	if rawBranch == "" {
		rawBranch = git.DetectBranch(root)
	}
	branch := ""
	if rawBranch != "" {
		branch = git.SanitizeBranch(rawBranch)
	}
	return &Store{Root: root, RawBranch: rawBranch, Branch: branch}, nil
}

// findExisting walks up from dir looking for a directory containing .incomm/.
func findExisting(dir string) (string, bool) {
	for {
		info, err := os.Stat(filepath.Join(dir, DirName))
		if err == nil && info.IsDir() {
			return dir, true
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return "", false
		}
		dir = parent
	}
}

// Dir returns the absolute path of the .incomm/ directory.
func (s *Store) Dir() string { return filepath.Join(s.Root, DirName) }

// NotesFileName returns the filename for the current branch. With a branch
// it is "notes_<branch>.json"; without one it is the legacy "notes.json".
func (s *Store) NotesFileName() string {
	if s.Branch == "" {
		return FileName
	}
	return "notes_" + s.Branch + ".json"
}

// NotesPath returns the absolute path of the branch-scoped notes file.
func (s *Store) NotesPath() string { return filepath.Join(s.Dir(), s.NotesFileName()) }

// Load reads the branch-scoped notes file. A missing file yields an empty,
// normalized NotesFile.
func (s *Store) Load() (*model.NotesFile, error) {
	data, err := os.ReadFile(s.NotesPath())
	if errors.Is(err, os.ErrNotExist) {
		return model.NewNotesFile(), nil
	}
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", s.NotesPath(), err)
	}
	var f model.NotesFile
	if err := json.Unmarshal(data, &f); err != nil {
		return nil, fmt.Errorf("parse %s: %w", s.NotesPath(), err)
	}
	f.Normalize()
	return &f, nil
}

// Save writes the branch-scoped notes file atomically (temp file + rename) with
// 2-space indent. The raw branch name is stamped into the JSON.
func (s *Store) Save(f *model.NotesFile) error {
	if s.RawBranch != "" {
		f.Branch = s.RawBranch
	}
	f.Normalize()
	data, err := json.MarshalIndent(f, "", "  ")
	if err != nil {
		return fmt.Errorf("encode notes: %w", err)
	}
	data = append(data, '\n')

	dir := s.Dir()
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("create %s: %w", dir, err)
	}
	tmp, err := os.CreateTemp(dir, ".notes-*.json.tmp")
	if err != nil {
		return fmt.Errorf("create temp file: %w", err)
	}
	tmpName := tmp.Name()
	defer os.Remove(tmpName) // no-op if the rename succeeded
	if _, err := tmp.Write(data); err != nil {
		tmp.Close()
		return fmt.Errorf("write temp file: %w", err)
	}
	if err := tmp.Sync(); err != nil {
		tmp.Close()
		return fmt.Errorf("sync temp file: %w", err)
	}
	if err := tmp.Close(); err != nil {
		return fmt.Errorf("close temp file: %w", err)
	}
	if err := os.Rename(tmpName, s.NotesPath()); err != nil {
		return fmt.Errorf("replace %s: %w", s.NotesFileName(), err)
	}
	return nil
}

// Clear deletes the branch-scoped notes file. It reports whether the file
// existed. The .incomm/ directory is removed too if it becomes empty.
func (s *Store) Clear() (bool, error) {
	err := os.Remove(s.NotesPath())
	if errors.Is(err, os.ErrNotExist) {
		return false, nil
	}
	if err != nil {
		return false, fmt.Errorf("remove %s: %w", s.NotesPath(), err)
	}
	// Best effort: drop the .incomm dir if empty.
	if entries, derr := os.ReadDir(s.Dir()); derr == nil && len(entries) == 0 {
		_ = os.Remove(s.Dir())
	}
	return true, nil
}

// RelFile converts a user-provided file path (absolute or relative to CWD) to a
// project-root-relative path using POSIX separators, as stored in the notes file.
func (s *Store) RelFile(path string) (string, error) {
	abs, err := filepath.Abs(path)
	if err != nil {
		return "", fmt.Errorf("resolve %q: %w", path, err)
	}
	rel, err := filepath.Rel(s.Root, abs)
	if err != nil {
		return "", fmt.Errorf("relativize %q against root %q: %w", path, s.Root, err)
	}
	if strings.HasPrefix(rel, ".."+string(filepath.Separator)) || rel == ".." {
		return "", fmt.Errorf("file %q is outside the project root %q", path, s.Root)
	}
	return filepath.ToSlash(rel), nil
}

// AbsFile resolves a project-root-relative note path back to an absolute path.
func (s *Store) AbsFile(rel string) string {
	return filepath.Join(s.Root, filepath.FromSlash(rel))
}

// ReadLines returns the lines of a project file (referenced by a note's rel
// path), with line breaks stripped. A trailing newline does not produce a
// trailing empty line.
func (s *Store) ReadLines(rel string) ([]string, error) {
	data, err := os.ReadFile(s.AbsFile(rel))
	if err != nil {
		return nil, err
	}
	return SplitLines(string(data)), nil
}

// SplitLines splits text into lines, normalizing CRLF/CR to LF semantics and
// dropping a single trailing newline so line counts match editor line numbers.
func SplitLines(text string) []string {
	text = strings.ReplaceAll(text, "\r\n", "\n")
	text = strings.ReplaceAll(text, "\r", "\n")
	if text == "" {
		return []string{}
	}
	if strings.HasSuffix(text, "\n") {
		text = text[:len(text)-1]
	}
	return strings.Split(text, "\n")
}

File: plugins/nvim/lua/incomm/store.lua
-- Disk IO for `.incomm/notes[_<branch>].json` -- a port of `cli/internal/store`.
--
-- Root resolution matches the CLI exactly: walk up from the starting directory
-- looking for an existing `.incomm/`, and fall back to the starting directory
-- itself (where `.incomm/` will be created on the first save). Writes are
-- atomic (temp file + rename) because the CLI and this plugin may both be
-- writing, and a reader must never see half a file.

local uv = vim.uv or vim.loop
local git = require("incomm.git")
local model = require("incomm.model")

local M = {}

M.DIR_NAME = ".incomm"
M.FILE_NAME = "notes.json" -- fallback when no branch is detected

---@class incomm.Store
---@field root string absolute path of the directory holding (or to hold) .incomm/
---@field raw_branch string raw git branch name, written into the JSON
---@field branch string filename-safe branch, "" for the legacy notes.json
local Store = {}
Store.__index = Store

--- Walk up from `dir` looking for a directory that contains `.incomm/`.
---@param dir string
---@return string?
local function find_existing(dir)
  while true do
    local stat = uv.fs_stat(dir .. "/" .. M.DIR_NAME)
    if stat and stat.type == "directory" then
      return dir
    end
    local parent = vim.fn.fnamemodify(dir, ":h")
    if parent == dir then
      return nil
    end
    dir = parent
  end
end

M.find_existing = find_existing

--- Resolve the project root and open a store for it.
---@param start_dir? string defaults to the current working directory
---@param explicit_branch? string skips auto-detection when given
---@return incomm.Store
function M.open(start_dir, explicit_branch)
  local base = vim.fn.fnamemodify(start_dir or uv.cwd() or ".", ":p")
  base = base:gsub("/+$", "")
  local root = find_existing(base) or base
  local raw_branch = explicit_branch or git.detect_branch(root)
  return setmetatable({
    root = root,
    raw_branch = raw_branch,
    branch = raw_branch ~= "" and git.sanitize_branch(raw_branch) or "",
  }, Store)
end

---@return string
function Store:dir()
  return self.root .. "/" .. M.DIR_NAME
end

--- `notes_<branch>.json`, or the legacy `notes.json` without a branch.
---@return string
function Store:notes_file_name()
  if self.branch == "" then
    return M.FILE_NAME
  end
  return "notes_" .. self.branch .. ".json"
end

---@return string
function Store:notes_path()
  return self:dir() .. "/" .. self:notes_file_name()
end

--- Raw contents of the notes file, or nil when it does not exist.
---@return string?
function Store:read_raw()
  return git.read_file(self:notes_path())
end

--- Load the branch-scoped notes file. A missing file yields an empty model.
---@return incomm.NotesFile, string? error
function Store:load()
  local data = self:read_raw()
  if not data then
    return model.new_file()
  end
  local parsed, err = model.decode(data)
  if not parsed then
    return model.new_file(), err
  end
  return parsed
end

--- Write the notes file atomically, stamping the raw branch name into the JSON.
--- Returns the exact bytes written, so the caller can recognise its own write
--- when the file watcher fires.
---@param f incomm.NotesFile
---@return string? written, string? error
function Store:save(f)
  if self.raw_branch ~= "" then
    f.branch = self.raw_branch
  end
  model.normalize(f)
  local data = model.encode(f)

  local dir = self:dir()
  if not uv.fs_stat(dir) then
    local ok = vim.fn.mkdir(dir, "p")
    if ok == 0 then
      return nil, "could not create " .. dir
    end
  end

  local tmp = string.format("%s/.notes-%d-%s.json.tmp", dir, uv.os_getpid(), model.new_id())
  local fd, open_err = uv.fs_open(tmp, "w", 420) -- 0644
  if not fd then
    return nil, open_err or ("could not open " .. tmp)
  end
  local ok, write_err = pcall(function()
    assert(uv.fs_write(fd, data, 0))
    uv.fs_fsync(fd)
  end)
  uv.fs_close(fd)
  if not ok then
    uv.fs_unlink(tmp)
    return nil, tostring(write_err)
  end
  local renamed, rename_err = uv.fs_rename(tmp, self:notes_path())
  if not renamed then
    uv.fs_unlink(tmp)
    return nil, rename_err or "rename failed"
  end
  return data
end

--- Delete the branch-scoped notes file, and `.incomm/` if it is left empty.
---@return boolean existed
function Store:clear()
  local path = self:notes_path()
  if not uv.fs_stat(path) then
    return false
  end
  uv.fs_unlink(path)
  local handle = uv.fs_scandir(self:dir())
  if handle and not uv.fs_scandir_next(handle) then
    uv.fs_rmdir(self:dir())
  end
  return true
end

--- Project-root-relative POSIX path for a file, or nil when it sits outside
--- the root (notes can only anchor inside the project).
---@param path string absolute or relative to the cwd
---@return string?
function Store:rel_file(path)
  local abs = vim.fn.fnamemodify(path, ":p")
  abs = abs:gsub("/+$", "")
  local root = self.root:gsub("/+$", "")
  if abs == root then
    return nil
  end
  if abs:sub(1, #root + 1) ~= root .. "/" then
    -- Resolve symlinks once before giving up: /tmp is a symlink on macOS, and
    -- a project opened through one would otherwise look like it is elsewhere.
    local real_abs = uv.fs_realpath(abs) or abs
    local real_root = uv.fs_realpath(root) or root
    if real_abs:sub(1, #real_root + 1) ~= real_root .. "/" then
      return nil
    end
    return real_abs:sub(#real_root + 2)
  end
  return abs:sub(#root + 2)
end

--- Absolute path of a note's project-relative file.
---@param rel string
---@return string
function Store:abs_file(rel)
  return self.root .. "/" .. rel
end

--- Lines of a project file as stored on disk, or nil when unreadable.
---@param rel string
---@return string[]?
function Store:read_lines(rel)
  local data = git.read_file(self:abs_file(rel))
  if not data then
    return nil
  end
  return require("incomm.anchor").split_lines(data)
end

return M
